README.md
# Front Matter [![Build Status](https://travis-ci.org/zhaocai/front-matter.rb.png?branch=master)](https://travis-ci.org/zhaocai/front-matter.rb)

* home  :: http://zhaocai.github.com/front-matter.rb/
* rdoc  :: http://rubydoc.info/gems/front-matter/
* code  :: https://github.com/zhaocai/front-matter.rb
* bugs  :: https://github.com/zhaocai/front-matter.rb/issues

## DESCRIPTION:

This gem extracts embeded info (yaml front matters, for example ) in source code
comments with predefined format:

1. the leading comment string can be different based on the language.
2. the leading column of yaml docs should be aligned to be converted to yaml.

### Format 1

      # ---
      # Gem           : front_matter
      # Author        : Zhao Cai
      # Email         : caizhaoff@gmail.com
      # Tag           : [ ruby, yaml, comment ]
      # ---

  OR

      # ------------- - ------------------------------------------------------------
      # Gem           : front_matter
      # Author        : Zhao Cai
      # Email         : caizhaoff@gmail.com
      # Homepage      : https://github.com/zhaocai/
      # Tag           : [ ruby, yaml, comment ]
      # ------------- - ------------------------------------------------------------

### Format 2
      # ---
      # Gem           : front_matter
      # Author        : Zhao Cai
      # Email         : caizhaoff@gmail.com
      # Tag           : [ ruby, yaml, comment ]
      # 

      ... source code


## SYNOPSIS:

```ruby
require 'front_matter'
require 'awesome_print'
require "yaml"

fm = FrontMatter.new(:unindent => true, :as_yaml => true)
file = "README.md"

front_matters = fm.extract_file(file)
ap YAML.load(front_matters[0])

# {
#   "Gem"      => "front_matter",
#   "Author"   => "Zhao Cai",
#   "Email"    => "caizhaoff@gmail.com",
#   "Tag"      => [
#     "ruby",
#     "yaml",
#     "comment"
#   ]
# }


```

## INSTALL:

`gem install front-matter`

## Requirement:

ruby 1.9+

## History

refer to [History.md]( https://raw.github.com/zhaocai/front-matter.rb/master/History.md )

## DEVELOPERS:

After checking out the source, run:

  $ rake newb

This task will install any missing dependencies, run the tests/specs,
and generate the RDoc.

## LICENSE:

(The MIT License)

Copyright (c) 2012, Zhao Cai

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md
# RELEASE HISTORY

## V1.2.1 / 2013-04-16

Simplify code by refining regex with grouped backreference (Zhao Cai <caizhaoff@gmail.com>)

Changes:

* 3 Major Enhancements

    * refine regex
    * rename project to front_matter
    * first working version.

* 3 Minor Enhancements

    * firstline..lastline option to search file in range
    * as_yaml option and spec test
    * add unindent option

* 1 Bug Fixes

    * join readlines


## V1.1.0 / 2013-04-16

Add options including :unindent, :as_yaml, :filetype, :firstline, :lastline (Zhao Cai <caizhaoff@gmail.com>)


## V1.0.1 / 2013-04-16

First Working Version (Zhao Cai <caizhaoff@gmail.com>)


## HEAD / 2013-04-17

Current Development (Zhao Cai)

Changes:

* 1 Major Enhancements

    * refine regex

* 2 Minor Enhancements

    * use vclog to update history
    * firstline..lastline option to search file in range

* 1 Bug Fixes

    * join readlines



spec/spec_helper.rb
$:.unshift File.join(File.dirname(__FILE__), *%w[.. lib])

require "rspec"

require 'facets/string'
require "front_matter"

RSpec.configure do |c|
  c.color_enabled = true

  # Use color not only in STDOUT but also in pagers and files
  c.tty = true

  c.formatter = :documentation # :progress, :html, :textmate

  c.mock_with :rspec
end



spec/front_matter_spec.rb
require "spec_helper"
require "yaml"
describe "Front_Matter" do
  before :all do

    test_dir = "test"
    Dir.mkdir(test_dir) unless Dir.exists?(test_dir)
    Dir.chdir(test_dir)
  end

  context "Extraction" do
    before :all do
      @fm =  FrontMatter.new
    end

    it "should extract yaml front matter" do
      code = %Q{
      # ---
      #       FileName : extract_sh
      #       Desc     : extract shell functions from shell script
      #       Author   : Zhao Cai <caizhaoff@gmail.com>
      # ---
      # 
      }.unindent

      valid_content = [
        "      FileName : extract_sh"                                ,
        "      Desc     : extract shell functions from shell script" ,
        "      Author   : Zhao Cai <caizhaoff@gmail.com>",
        ""
      ]
      front_matters = @fm.extract(code)

      front_matters[0].should == valid_content.join("\n")
    end

    it "should extract header front matter" do
      code = %Q{
      # --------------- ------------------------------------------------------------
      #       FileName : extract_sh
      #       Desc     : extract shell functions from shell script
      #       Author   : Zhao Cai <caizhaoff@gmail.com>
      # --------------- ------------------------------------------------------------
      }.unindent

      valid_content = [
        "      FileName : extract_sh"                                ,
        "      Desc     : extract shell functions from shell script" ,
        "      Author   : Zhao Cai <caizhaoff@gmail.com>",
        ""
      ]
      front_matters = @fm.extract(code)

      front_matters[0].should == valid_content.join("\n")

    end

    it "should properly extract yaml front matter when there is no ending" do
      code = %Q{
      # ---
      #       FileName : extract_sh
      #       Desc     : extract shell functions from shell script
      #       Author   : Zhao Cai <caizhaoff@gmail.com>

      abc
      }.unindent

      valid_content = [
        "      FileName : extract_sh"                                ,
        "      Desc     : extract shell functions from shell script" ,
        "      Author   : Zhao Cai <caizhaoff@gmail.com>",
        ""
      ]
      front_matters = @fm.extract(code)

      front_matters[0].should == valid_content.join("\n")
    end
  end



  context "Unindent" do
    before :all do
      @fm =  FrontMatter.new(:unindent => true)
    end

    it "should extract yaml front matter" do
      code = %Q{
      # ---
      #       FileName : extract_sh
      #       Desc     : extract shell functions from shell script
      #       Author   : Zhao Cai <caizhaoff@gmail.com>
      # ---
      # 
      }.unindent

      valid_content = [
        "FileName : extract_sh"                                ,
        "Desc     : extract shell functions from shell script" ,
        "Author   : Zhao Cai <caizhaoff@gmail.com>",
        ""
      ]
      front_matters = @fm.extract(code)

      front_matters[0].should == valid_content.join("\n")

    end

    it "should extract header front matter" do
      code = %Q{
      # --------------- ------------------------------------------------------------
      #       FileName : extract_sh
      #       Desc     : extract shell functions from shell script
      #       Author   : Zhao Cai <caizhaoff@gmail.com>
      # --------------- ------------------------------------------------------------
      }.unindent

      valid_content = [
        "FileName : extract_sh"                                ,
        "Desc     : extract shell functions from shell script" ,
        "Author   : Zhao Cai <caizhaoff@gmail.com>",
        ""
      ]
      front_matters = @fm.extract(code)

      front_matters[0].should == valid_content.join("\n")

    end

    it "should properly extract yaml front matter when there is no ending" do
      code = %Q{
      # ---
      #       FileName : extract_sh
      #       Desc     : extract shell functions from shell script
      #       Author   : Zhao Cai <caizhaoff@gmail.com>

      abc
      }.unindent

      valid_content = [
        "FileName : extract_sh"                                ,
        "Desc     : extract shell functions from shell script" ,
        "Author   : Zhao Cai <caizhaoff@gmail.com>",
        ""
      ]
      front_matters = @fm.extract(code)

      front_matters[0].should == valid_content.join("\n")
    end
  end

  context "YAML" do
    before :all do
      @fm =  FrontMatter.new(:unindent => true, :as_yaml => true)
    end

    it "should extract and convert to valid yaml" do
      code = %Q{
      # ---
      #       FileName : extract_sh
      #       Desc     : extract shell functions from shell script
      #       Author   : Zhao Cai <caizhaoff@gmail.com>
      # ---
      # 
      }.unindent

      valid_yaml = YAML.load([
        "---",
        "FileName : extract_sh"                                ,
        "Desc     : extract shell functions from shell script" ,
        "Author   : Zhao Cai <caizhaoff@gmail.com>"
      ].join("\n"))

      front_matters = @fm.extract(code)

      YAML.load(front_matters[0]).should == valid_yaml
    end

  end
end





test/test_extr_comment.rb
require "test/unit"
require "extr_comment"
require 'awesome_print'


class TestExtrComment < Test::Unit::TestCase

  def test_extract_header_style
    code= <<"__RUBY__"
      # --------------- ------------------------------------------------------------
      #       FileName : extract_sh
      #           Desc : extract shell functions from shell script
      #         Author : Zhao Cai <caizhaoff@gmail.com>
      # --------------- ------------------------------------------------------------
__RUBY__

    valid_content = [
      "FileName : extract_sh",
      "Desc : extract shell functions from shell script",
      "Author : Zhao Cai <caizhaoff@gmail.com>"
    ]
    ec = ExtrComment.new()
    content = ec.extract_lines(code.split("\n").map(&:chomp))

    assert_nil(content[:default])
    assert_not_nil(content[:header])
    assert_equal(content[:header][0][:valid], valid_content)

  end

  def test_extract_default_style
    code= <<"__RUBY__"
      # ---
      #       FileName : extract_sh
      #           Desc : extract shell functions from shell script
      #         Author : Zhao Cai <caizhaoff@gmail.com>
      # ...
__RUBY__

    valid_content = [
      "FileName : extract_sh",
      "Desc : extract shell functions from shell script",
      "Author : Zhao Cai <caizhaoff@gmail.com>"
    ]
    ec = ExtrComment.new()
    content = ec.extract_lines(code.split("\n").map(&:chomp))

    assert_not_nil(content[:default])
    assert_nil(content[:header])
    assert_equal(content[:default][0][:valid], valid_content)
  end

  def test_no_ending
    code= <<"__RUBY__"
      # ---
      #       FileName : extract_sh
      #           Desc : extract shell functions from shell script
      #         Author : Zhao Cai <caizhaoff@gmail.com>

abc
__RUBY__

    valid_content = [
      "FileName : extract_sh",
      "Desc : extract shell functions from shell script",
      "Author : Zhao Cai <caizhaoff@gmail.com>"
    ]
    ec = ExtrComment.new()
    content = ec.extract_lines(code.split("\n").map(&:chomp))

    assert_not_nil(content[:default])
    assert_nil(content[:header])
    assert_equal(content[:default][0][:unbound], valid_content)
  end

end


lib/extr_comment.rb
#!/usr/bin/env ruby
# =============== ============================================================
#  FileName      : extr_comment
#  Desc          : extract comment from source code
#  Author        : Zhao Cai <caizhaoff@gmail.com>
#  HomePage      : https://github.com/zhaocai/extr_comment.rb
#  Version       : 0.1
#  Date Created  : Sat 25 Aug 2012 01:36:13 AM EDT
#  Last Modified : Fri 14 Sep 2012 09:05:30 PM EDT
#  Tag           : [ ruby, comment, extract ]
#  Copyright     : (c) 2012 by Zhao Cai,
#                  Released under current GPL license.
# =============== ============================================================

# [TODO]( align ) @zhaocai @start(2012-08-25 05:54)
class ExtrComment
  VERSION = '1.0.1'
  attr_accessor :patterns
  def initialize(patterns={} )
    cmarker   = %r{(?<comment> ^\s* \W{1,2} \s*)}x
    @patterns = {
      :header => {
        :filetype => %r{.*},
        :cmarker  => cmarker,
        :start    => %r{#{cmarker} (?<start> [-=]{3,} \s [-=]{3,}$) }x ,
        :content  => %r{#{cmarker} (?<content> .* $)          }x ,
        :end      => %r{#{cmarker} (?<end> [-=]{3,} \s [-=]{3,}$)   }x ,
      },
      :default => {
        :filetype => %r{.*},
        :cmarker  => cmarker,
        :start    => %r{#{cmarker} (?<start> \s*  -{3} $) }x ,
        :content  => %r{#{cmarker} (?<content> .* $)      }x ,
        :end      => %r{#{cmarker} (?<end> \s*  \.{3} $)  }x ,
      },
    }

    @patterns.merge!(patterns)

  end

  def extract_lines(lines, filetype=[])
    content={}
    unless filetype.empty?
      patterns = @patterns.select { |k,v|
        filetype.any { |ft| v[:filetype].match(ft)}
      }
    else
      patterns = @patterns
    end

    patterns.each { |k, v|
      content[k] = []
      in_comment = false
      in_content = {
        :valid   => [],
        :invalid => [],
        :unbound => [],
      }

      lines.each { |line|
        if ! in_comment
          case line
          when v[:start]
            in_comment = true
            next
          end
        else
          if v[:end] =~ line
            in_comment = false
            content[k].push(in_content)
            next
          end

          m = v[:content].match(line)
          begin
            in_content[:valid].push(m[:content])
          rescue IndexError,NoMethodError
            in_content[:invalid].push(line)
          end
        end
      }

      # fail to match ending
      if in_comment
        in_content[:unbound] = in_content[:valid]
        in_content[:valid] = []
        content[k].push(in_content)
      end
    }

    return content.delete_if {|k, v| v.empty?}
  end

  def extract_file(infile, filetype=[])
    return self.extract_lines(File.readlines(infile).map(&:chomp), filetype)
  end

end





lib/front_matter.rb
#!/usr/bin/env ruby
# =============== ============================================================
#  Desc          : extract (YAML) front matter from source code
#  Author        : Zhao Cai <caizhaoff@gmail.com>
#  HomePage      : https://github.com/zhaocai/front_matter.rb
#  Date Created  : Sat 25 Aug 2012 01:36:13 AM EDT
#  Last Modified : Sun 14 Apr 2013 08:05:00 PM EDT
#  Tag           : [ ruby, YAML, comment, extract ]
#  Copyright     : (c) 2012 by Zhao Cai,
#                  Released under current GPL license.
# =============== ============================================================

require 'facets/hash'
require 'facets/string'
require 'front_matter/core/array'

## 
# Extracts embeded info (yaml front matters, for example ) in source
# code comments with predefined format.

class FrontMatter
  VERSION = '1.2.1'

  ##
  # *Optional*: setting for FrontMatter
  #
  # - :patterns => add new patterns to extract 
  # - :unindent => unindent leading whitespace  
  # - :as_yaml  => results are formated for YAML.load
  #
  attr_accessor :setting

  def initialize( opts={} )
    comment_marker   = %r{(?<comment> ^\s* \W{1,2} )}x
    @setting = {
      :patterns => {
        :header => {
          :filetype => %r{.*},
          :pattern  => %r{
            (?<start> #{comment_marker} \s* [-=#*]{3,} .* )[\r\n]
            (?<content> (?: \k<comment> .* [\r\n])+)
            \k<start>
            }x
        },
        :yaml => {
          :filetype => %r{.*},
          :pattern  => %r{
            (?<start> #{comment_marker} \s* -{3} )[\r\n]
            (?<content> (?: \k<comment> .* [\r\n])+)
            (?<empty>^ \s* $)
            }x
        },
      },
      :unindent   => false ,
      :as_yaml    => false ,
    }
    @setting.merge!(opts)
  end

  ##
  # Extract from string
  # - contents : String
  # - filetype : limit patterns to Array of filetypes, default []
  # 
  # *Returns*
  # - Array of extracted front matter, comment string are removed
  #
  # *Raises*
  # - None from this class
  #
  def extract(contents, filetype=[])
    unless filetype.empty?
      patterns = @setting[:patterns].select { |kind, pattern|
        filetype.any { |ft| pattern[:filetype].match(ft)}
      }
    else
      patterns = @setting[:patterns]
    end

    union_patterns = Regexp.union(patterns.collect{ |k, p| p[:pattern] })
    results = [] 
    contents.mscan(union_patterns).each { |m| 
      results << m[:content].gsub(/^#{Regexp.escape(m[:comment])}/, "") 
    }
    
    results.map! { |r| r.unindent } if @setting[:unindent]
    results.map! { |r| "---\n#{r}" } if @setting[:as_yaml]

    return results
  end


  ##
  # Extract from file
  #
  # - path : file path
  # - opts : 
  #   - :filetype  : Array of filetypes
  #   - :firstline
  #   - :lastline
  # 
  # *Returns*
  # - Array of extracted front matter, comment string are removed
  #
  # *Raises*
  # - None from this class
  #
  # *Example*
  #   fm = FrontMatter.new(:unindent => true, :as_yaml => true)
  #   puts fm.extract_file(__FILE__, :firstline => 1, :lastline => 15)
  #
  # The output would be:
  #   ---
  #   Desc          : extract (YAML) front matter from source code
  #   Author        : Zhao Cai <caizhaoff@gmail.com>
  #   HomePage      : https://github.com/zhaocai/front_matter.rb
  #   Date Created  : Sat 25 Aug 2012 01:36:13 AM EDT
  #   Last Modified : Sun 14 Apr 2013 08:05:00 PM EDT
  #   Tag           : [ ruby, YAML, comment, extract ]
  #   Copyright     : (c) 2012 by Zhao Cai,
  #                     Released under current GPL license.

  def extract_file(path, opts={})

    filetype = opts[:filetype] ? opts[:filetype] : []
    if opts[:firstline] || opts[:lastline]
      firstline = opts[:firstline] ? opts[:firstline] - 1 : 0
      lastline = opts[:lastline] ? opts[:lastline] -1 : -1
      return extract(File.readlines(path)[firstline..lastline].join, filetype)
    else
      return extract(File.read(path), filetype)
    end

  end

end


if __FILE__ == $0
  fm = FrontMatter.new(:unindent => true, :as_yaml => true)
  puts fm.extract_file(__FILE__, :firstline => 1, :lastline => 15)
end
